ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java

```diff
@@ -419,115 +402,28 @@ public boolean hasSucceeded() {
-      //        while TTLUpdate/Delete don't. Refer to routerOperationTrackerIncludeDownReplicas related code.
-      //        So it has tried the OFFLINE replicas already.
-      //    b). But with the recent Helix implementation, when the server is crashed,
-      //        the helix external view sometimes doesn't include the replica at all, not even in the OFFLINE category.
-      //        So we check how many replicas from the originating DC we read already.
-      //        Since we put routerPutSuccessTarget copies in the originating DC,
-      //        we should at least read "originatingDcTotalReplicaCount - routerConfig.routerPutSuccessTarget + 1" replicas
-      //        to make sure we read one of the copies.
-      //        If the eligible originating replica count is less that that, we should retry.
-      if (hasFailedOnCrossColoNotFound()
-          && (allReplicaCount - totalNotFoundCount >= replicaSuccessTarget)
-          && (originatingDcEligibleReplicaCount < originatingDcTotalReplicaCount - routerConfig.routerPutSuccessTarget + 1)) {
-        logger.info(
-            "Terminating {} on {} due to disk down count and total Not_Found count from eligible replicas and some "
-                + "other replicas being unavailable. CrossColoEnabled: {}, DiskDownCount: {}, TotalNotFoundCount: {}, "
-                + "TotalReplicaCount: {}, replicaSuccessTarget: {}, OfflineReplicaCount: {}, allReplicaCount: {}, "
-                + "originatingDcEligibleReplicaCount {} {} "
-                + "replicasByState = {}",
-            routerOperation, partitionId, crossColoEnabled, diskDownCount, totalNotFoundCount, totalReplicaCount,
-            replicaSuccessTarget, totalOfflineReplicaCount, allReplicaCount, originatingDcEligibleReplicaCount,
-            getBlobIdLog(), allDcReplicasByState);
-        routerMetrics.failedMaybeDueToTotalOfflineReplicasCount.inc();
-        return true;
-      }
-      return false;
+    if (originatingDcNotFoundFailureThreshold == NOT_FOUND_THRESHOLD_INVALID_VALUE) {
+      // Since some originating DC replicas are down or in bootstrap state, we can't conclude if blob is missing.
+      return FailedOnNotFoundResult.INCONCLUSIVE;
     }
 
-    // For Delete and TtlUpdate
-    // We mark the failure as due to offline replicas when we know that we couldn't find the blob in eligible replicas,
-    // and remaining replicas are enough to make the request successful, and there is atleast one offline replica in the
-    // remaining set of replicas. The offline replicas can come back up in future to make the request successful, and
-    // hence such a failure should be deemed as retryable.
-    if (hasFailedOnOriginatingDcNotFound()
-        && originatingDcTotalReplicaCount - originatingDcNotFoundCount >= replicaSuccessTarget
-        && originatingDcOfflineReplicaCount > 0) {
+    if (originatingDcNotFoundCount >= originatingDcNotFoundFailureThreshold) {
```

Comment:
let's add another condition before this, as long as there is a success response, we return False here.
There is some corner case here.
If we don't check all replicas in originating dc, then the threshold here is 2.
When we upload a blob to originating DC, assume there is only replica 1 and 2 get the blob and 3 didn't get it. And when we try to delete this blob, 2 become bootstrap. Now we have 1 as leader, 2 as bootstrap and 3 as standby. We have 2 replicas as leader or standby so the threshold is not -1. However we will get not found from 2 and 3. This way we would return NOT_FOUND,even when we get success from 1.

Reply:
or we should always check all the replicas in originating DC but this way, we would return more 503 than we should.

Reply:
Good point. Thank you for mentioning the case. Updated the code.

Reply:
I think probably Justin brought a generic question which always exist, 
In the checkAndMaybeComplete function, for NOT_FOUND, if hasFailedOnNotFound is true, we set the exception to NOT_FOUND no matter what the previous responses we ever received are. 

```
if (operationTracker.hasFailedOnNotFound()) {
    operationException.set(BlobDoesNotExist))
}
```

In Justin's case, replica 1 returns success status. 
What if replica 1 returns BlobAuthorizationFailure/BlobExpired etc? We are supposed to return these high precedence errors but we return NOT_FOUND.

Probably something like

```
if (operationTracker.hasFailedOnNotFound() && no_exception_received_higher_than_AmbryUnavailable) {
    operationException.set(BlobDoesNotExist))
}
```